Assets/Shopify/UIToolkit/Editor/ProductPicker.cs

```diff
@@ -0,0 +1,168 @@
+﻿namespace Shopify.UIToolkit.Editor {
+    using System.Collections;
+    using System.Collections.Generic;
+    using UnityEngine;
+    using UnityEditor;
+    using Shopify.Unity;
+    using Shopify.Unity.SDK;
+    using System;
+    using System.Linq;
+
+    public interface IProductPickerView {
+        void DrawProductLoadingState(SerializedObject serializedObject);
+        void DrawProductPicker(SerializedObject serializedObject, Dictionary<string, string> productNameToGIDMap);
+        void DrawProductLoadingError(SerializedObject serializedObject);
+        void DrawShopHasNoProductsState(SerializedObject serializedObject);
+    }
+
+    public interface IProductPickerController {
+        void OnShouldRefreshProductList();
+    }
+
+    public class ProductPicker : IProductPickerController {
+        private IProductPickerView _view;
+        private static Dictionary<string, string> _cachedProductNameToGIDMap;
+        private Dictionary<string, string> _productNameToGIDMap;
+        private ShopifyClient _client;
+        private ShopifyError _error;
+
+        public bool ProductsFinishedLoading() {
+            return _productNameToGIDMap != null || _error != null;
+        }
+
+        public ProductPicker(ShopifyClient client, IProductPickerView view = null) {
+            _view = view ?? new ProductPickerView(this);
+            _client = client;
+            LoadProducts();
+        }
+
+        public void DrawInspectorGUI(SerializedObject serializedObject) {
+            if (_error != null) {
+                _view.DrawProductLoadingError(serializedObject);
+                return;
+            }
+
+            if (!ProductsFinishedLoading()) {
+                _view.DrawProductLoadingState(serializedObject);
+                return;
+            }
+
+            if (_productNameToGIDMap.Count == 0) {
+                _view.DrawShopHasNoProductsState(serializedObject);
+                return;
+            }
+
+            _view.DrawProductPicker(serializedObject, _productNameToGIDMap);
+        }
+
+        public void LoadProducts() {
+            if (_cachedProductNameToGIDMap != null) {
+                _productNameToGIDMap = _cachedProductNameToGIDMap;
```

Comment:
What is this cache for? We seem to set it in a few places, but I don't see it being read anywhere..?

Reply:
its being read right here :) its written to after the products have loaded once. The reason for this is we get a new editor instance everytime that we select the object, so this cache will only be cleared after a recompilation instead of the product list reloading every time we look at the object. Its used mostly for an optimistic load, where we are showing the last results while still fetching in the background to see if anything changed

Reply:
Doesn't `_productNameToGIDMap =` get called at https://github.com/Shopify/unity-buy-sdk/pull/400/files#diff-181c13ea5c2cfc8a4f52c91322bb3d54R102 anyways, once the recursive page load finishes?

Reply:
yep, the idea is that while the page is loading you can still browse the previous results. Especially cause 99% of the time those results will be the same as the results you have.